src/network.rs

```diff
@@ -30,23 +30,53 @@ pub trait Network {
     type Block: Block<Payload = Self::Payload>;
     type Message: Message<ValidatorId = Self::ValidatorId>;
     type Payload;
-    type Round: Round;
     type ValidatorId: ValidatorId;
+    type ValidatorIdRef: AsRef<Self::ValidatorId>;
+    type ValidatorIdSorted: Iterator<Item = Self::ValidatorIdRef>;
 
     fn broadcast(&mut self, message: &Self::Message);
-    fn increment_round(round: Self::Round) -> Self::Round;
-    fn is_validator(&self, round: Self::Round, validator: &Self::ValidatorId) -> bool;
-    fn validators(&self, round: Self::Round) -> usize;
-    fn leader(&self, round: Self::Round) -> Option<&Self::ValidatorId>;
+    fn validators_sorted(&self, round: &HeightRound) -> Self::ValidatorIdSorted;
 
     /// Generate the block payload to allow the creation of a new block.
     fn block_payload(&self) -> Self::Payload;
 
+    fn leader(&self, round: &HeightRound) -> Option<Self::ValidatorIdRef> {
+        let count = self.validators_count(round) as u64;
+
+        (count > 0)
+            .then(|| {
+                let h = round.height();
+                let r = round.round();
+
+                let index = (h + r) % count;
```

Comment:
This isn't a very fair balanced round-robin, but since it's easily overridable it's not a big deal. We could override this to integrate with a separate staking module to do roulette wheel selection or some other kind of stake-weighted round robin.